### PR TITLE
Add shift resource bindings in Vulkan

### LINKED ISSUE #37: Vulkan descriptor shifting
**Is your feature request related to a problem? Please describe.**
Its not a problem but probably good improvements when porting shaders to vulkan backend.

**Describe the solution you'd like**
Add the missing -fvk-{b|s|t|u}-shift options (https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/SPIR-V.rst).

**Describe alternatives you've considered**
None

**Additional context**
None

### LINKED ISSUE #41: Options::packMatricesInRowMajor is inverted
**Describe the bug**
Options::packMatricesInRowMajor is inverted. Setting it to true, passes in '-Zpc' which makes the compiler generate column major matrices.

**To Reproduce**
Inspect the source code:
```
        // HLSL matrices are translated into SPIR-V OpTypeMatrixs in a transposed manner,
        // See also https://antiagainst.github.io/post/hlsl-for-vulkan-matrices/
        if (options.packMatricesInRowMajor)
        {
            dxcArgStrings.push_back(L"-Zpc");
        }
        else
        {
            dxcArgStrings.push_back(L"-Zpr");
        }
```

**Expected behavior**
The values need to be inverted. Reading the comment it seems that this might've been intentionally inverted to fix SPIRV codegen, but my DX12/Vulkan/GLES engine works correctly for all three APIs when I pass in -Zpr and my matrices are row major in memory.

### PR DESCRIPTION
**Pull request by Jorgemagic**
Hi @gongminmin ,

I added 4 new parameters in Options struct:

int shiftAllTexturesBindings: First binding number for Vulkan textures.
int shiftAllSamplersBindings: First binding number for Vulkan samplers.
int shiftAllCBuffersBindings: First binding number for Vulkan constant buffers.
int shiftAllUABuffersBindings: First binding number for Vulkan unordered access buffers.

It is equivalent to use -fvk-{b|s|t|u}-shift in DXC command-line. That is a very important option when you compile HLSL to SPIRV [HLSL register and Vulkan binding](https://github.com/microsoft/DirectXShaderCompiler/blob/master/docs/SPIR-V.rst#hlsl-register-and-vulkan-binding).

Finally, I inverted packMatricesInRowMajor option to fix the issue.

Resolve #37 
Resolve #41

### CODE REVIEW

**Comment by gongminmin** [comment]
I restart the CI but it's still fail. The problem is you need to update the expected test results under Source\Tests\Data\Expected to reflect your changes to -Zpr.

File: Source/Tests/Data/Expected/Transform_VS.msl [hunk 1]
```diff
@@ -21,7 +21,7 @@ struct main0_in
 vertex main0_out main0(main0_in in [[stage_in]], constant type_cbVS& cbVS [[buffer(0)]])
 {
     main0_out out = {};
-    out.gl_Position = in.in_var_POSITION * cbVS.wvp;
+    out.gl_Position = cbVS.wvp * in.in_var_POSITION;
```
Comment (gongminmin) [inline Source/Tests/Data/Expected/Transform_VS.msl]: I still have a little concern about the matrix. My intention is on CPU side, the caller need to transpose a matrix before putting it to cbuffer, and writing the shader in mul(vec, matrix) way. It gets consistent with D3D and HLSL. My engine hasn't get that far to enable to test the translated GLSL. Could you explain me a little more about how to use the matrix in CPU side? Thanks.

Comment (Jorgemagic) [reply Source/Tests/Data/Expected/Transform_VS.msl]: Hi @gongminmin,

I recommend you check out Fabian Giesen's [post ](http://www.kiransprojects.com/blog/2014/row-vs-column/)about row-major and column-major matrix layouts and this Lei Zhang [post ](https://antiagainst.github.io/post/hlsl-for-vulkan-matrices/) about how to translate HLSL matrices to Vulkan.

Column-major and Row major indicate how the matrix is stored in memory, for example
a11 a12 a13
a21 a22 a23
a31 a32 a33

**Row major** is stored in memory as a11, a12, a13, a21, a22, a23, a31, a32, a33
**Column-major** is stored in memory as a11, a21, a31, a12, a22, a32, a13, a23, a33

If you want to convert from a layout to other, only you need to transpose the matrix.

Then [HLSL ](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-per-component-math#matrix-ordering) use column-major by default, but GLSL use row-major by default.

DirectX compiler allows you specify:
**-Zpc**: Packs matrices in column-major order by default
**-Zpr**: Packs matrices in row-major order by default

I working on [WaveEngine](https://waveengine.net/) and we use HLSL as main shader languages so we store matrices as row-major in CPU and send its directly to DirectX without transpose but use -Zpr to specify row-major by default in the DirectX compiler (dxc), then we have:

**HLSL**
```
cbuffer Matrices : register(b0)
{
	float4x4 worldViewProj;
};
```

**GLSL**
```
layout(binding = 0, std140) uniform type_Matrices
{
    mat4 worldViewProj;
} Matrices;
```

The layout doesn't need to be specified, either you're using HLSL or GLSL. 

Note. You can try WaveEngine Effect Editor to convert automatically your HLSL shader to other languages (taking advantage of ShaderConductor library)

Comment (gongminmin) [reply Source/Tests/Data/Expected/Transform_VS.msl]: Got it. Thanks very much for your explanation.

**Pull request merged**